### PR TITLE
Rewrite front end to work using AngularJS.

### PR DESCRIPTION
**Pull request by ganewman**
### What is a quick description of the change?

These changes establish a file structure and organization for the front end of the project using AngularJS modules and components.
The front end is now split into 4 modules, with 4 components:
- The `navigation` module, which contains the `header` and `navBar` components.
- The `directoryPage` module, which contains the `directoryPage` component.
- The `candidatePage` module, which contains the `candidatePage` component. The previous commit had 3 separate components within this module, but they have been combined to avoid having to retrieve data from the database more than once.

These three modules are contained within the larger `voterCentral` module, which encompasses the entire AngularJS app. All components except the `navBar` and `header` components also have corresponding template files (the templates for these two are embedded).

This code can now be automatically merged, as it works with the most recent version on master.

### Is this fixing an issue?

none

### Are there more details that are relevant?

none

### Check lists (check `x` in `[ ]` of list items)

- [  ] Test written/updating
- [X] Tests passing
- [X] Coding style (indentation, etc)

Please explain why any are not present, if any.

### Any additional comments?

{Fill in here or remove (leave question)}

### CODE REVIEW

File: project/src/main/webapp/app.module.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// use modern JavaScript (ES5)
+"use strict"
+
+/**
+ * Adds a official election/candidate information (brief) to the homepage.
```
Comment (suparnabarua) [inline project/src/main/webapp/app.module.js]: nit. Rewrite this comment. Is the election/candidate official or the information official?

File: project/src/main/webapp/candidate-page/official-info.template.html [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<h5>This is some placeholder text! The full HTML for the official candidate
+    information will eventually go here!</h5>
+
```
Comment (kellyktan) [inline project/src/main/webapp/candidate-page/official-info.template.html]: end of file newline?

**Review by kellyktan — COMMENTED** [review]
image

So maybe I'm doing something dumb but this is what I get when running the server locally with `mvn package appengine:run` and input a zip code... help

File: project/src/main/webapp/candidate-page/official-info.template.html (hunk 2 above)
Comment (ganewman) [reply project/src/main/webapp/candidate-page/official-info.template.html]: I'm confused. Doesn't it have the newline?

**Comment by ganewman** [comment]
I pushed a fix to the form on the index page which should change how things redirect to the directory page. Let me know if everything is still breaking... it all works for me locally so that's pretty concerning.

File: project/src/main/webapp/navigation/header.component.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+// use modern JavaScript (ES5)
+"use strict"
+
+angular.module('navigation').component('header', {
+  template: `<h1>Voter Central</h1>`,
+  controller: function generateHeader() {
+    // Generate any dynamic info for header
+  }
+});
```
Comment (suparnabarua) [inline project/src/main/webapp/navigation/header.component.js]: nit. add new line at end of file

File: project/src/main/webapp/navigation/nav-bar.component.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    <br>
+    <a href="/education.html">Voter Education</a>
+    <br>
+    <a href="/">Polling Locations</a>
+    <br>`,
+  controller: function generateNavBar() {
+    // Dynamically generate links and pages relating to directory page etc.
+  
+  }
+});
```
Comment (suparnabarua) [inline project/src/main/webapp/navigation/nav-bar.component.js]: nit. add new line at end of file

**Review by suparnabarua — CHANGES_REQUESTED** [review]

File: project/src/main/webapp/app.module.js (hunk 1 above)
Comment (ganewman) [reply project/src/main/webapp/app.module.js]: Honestly I'm not sure. I think this is an older comment. I'll fix it now!

File: project/src/main/webapp/navigation/header.component.js (hunk 3 above)
Comment (ganewman) [reply project/src/main/webapp/navigation/header.component.js]: Aw jeez. I really thought I caught all of these this time.

File: project/src/main/webapp/app.module.js (hunk 1 above)
Comment (ganewman) [reply project/src/main/webapp/app.module.js]: Oh, actually, that whole function was just a weird placeholder and doesn't go in that file. I just got rid of it.

**Review by KathyFeiyang — COMMENTED** [review]
I believe `node_modules/` is conventionally put into .gitignore.

**Comment by ganewman** [comment]
Oh, I didn't know that. Wouldn't that mean you all have to download it locally? It's pretty easy, I can link a tutorial.

File: project/src/main/webapp/app.module.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  // Unpack response.
+
+  // Add official election/candidate information to HTML.
+
+  // Add URLs, that send GET request to /candidate with candidate ID, to HTML.
+
+}
+
+angular.module('voterCentral',
+    ['navigation', 'directoryPage', 'candidatePage']);
```
Comment (KathyFeiyang) [inline project/src/main/webapp/app.module.js:19]: Have you decided whether to include the `eduationPage`?

File: project/src/main/webapp/directory-page/elections-table.template.html [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+              <td>Republican Party</td>
+              <td>No</td>
+            </tr>
+          </table>
+        </li>
+      </ul>
+    </div>
+  </div>
+  <hr>
+</div>
```
Comment (KathyFeiyang) [inline project/src/main/webapp/directory-page/elections-table.template.html]: nit: add new line

File: project/src/main/webapp/navigation/nav-bar.component.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+// use modern JavaScript (ES5)
+"use strict"
+
+angular.module('navigation').component('navBar', {
+  template: `<a href="/index.html">Home</a>
+    <br>
+    <a href="/">Resources</a>
+    <br>
+    <a href="/">Polling Locations</a>
+    <br>`,
```
Comment (KathyFeiyang) [inline project/src/main/webapp/navigation/nav-bar.component.js]: nit: indent 4+ spaces when you wrap lines

File: project/src/main/webapp/navigation/navigation.module.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// use modern JavaScript (ES5)
+"use strict"
+
+angular.module('navigation', []);
```
Comment (KathyFeiyang) [inline project/src/main/webapp/navigation/navigation.module.js]: nit: add new line at EOF

File: project/src/main/webapp/style.css [hunk 9]
```diff
@@ -8,6 +8,11 @@ body {
   text-align: center;
```
Comment (KathyFeiyang) [inline project/src/main/webapp/style.css:8]: nit: rules are usually ordered alphabetically

**Review by KathyFeiyang — CHANGES_REQUESTED** [review]

**Review by kellyktan — CHANGES_REQUESTED** [review]
Can you add the "download angular" instructions/command to the relevant readme?

File: project/src/main/webapp/app.module.js (hunk 5 above)
Comment (ganewman) [reply project/src/main/webapp/app.module.js:19]: Either way, it shouldn't need a corresponding module, as it's static and doesn't need a component. Theoretically if I get the routing working in the future, I could turn it from a regular HTML file into a template. But it works as is.

**Review by suparnabarua — APPROVED** [review]
LGTM pending resolution of open comments

**Comment by ganewman** [comment]
> Can you add the "download angular" instructions/command to the relevant readme?

Should be done now!

File: README.md [hunk 10]
```diff
@@ -4,3 +4,10 @@ This repo contains the STEP capstone project: "Voter Central".
 
 * Live server: [go/voter-central](http://go/voter-central)
 * Design doc: [go/voter-central-dd](http://go/voter-central-dd)
+
+The project requires the use of AngularJS. To install, make sure you have
+a recent version of `npm` installed, then run this command:
+
+```bash
+npm install angular
+```
```
Comment (kellyktan) [inline README.md]: Maybe move this instruction to the /project directory and say something like "Under `src/main/webapp/` run `npm install angular`".  Discovered the problems I was having with deploying this locally was that the webapp doesn't really "see" anything under the `webapp` directory.  Once that's done, you don't have to use relative paths for importing angular.  Can just use `/node_modules/angular/angular.js` in the html files.

**Review by kellyktan — APPROVED** [review]
Looks great, thanks!

File: project/src/main/webapp/app.module.js (hunk 5 above)
Comment (KathyFeiyang) [reply project/src/main/webapp/app.module.js:19]: Sounds great!

**Review by KathyFeiyang — APPROVED** [review]
LGTM! pending the removal of duplicated `package-lock.json` and `package.json`

**Review by tjyoung99 — APPROVED** [review]

**Pull request merged**